### PR TITLE
use paddle.to_tensor instead of blocking_queue

### PR DESCRIPTION
**Pull request by heavengate**
### PR types
Function optimization

### PR changes
APIs

### Describe

use `paddle.to_tensor(..., place=paddle.CUDAPinnedPlace())` instead of `py_reader + blocking_queue` to pin memory for following improvements:
1. support data in `dict` `list` `str` format: blocking queue need numpy array data format as input, which limit Dataset can only yield data as a list of numpy array
2. make DataLoader exiting more robust for blocking queue hold resources(threads, memory) and may raise error before python exit
3. make `__del__` can be called when iteration breaks so that worker process can safely exit when iteration breaks before stop iteration(error may also occur in this situation)
4. simplify data pipeline from `numpy array -> cpu tensor ->  pinned/gpu tensor` to  `numpy array -> pinned/gpu tensor`
5. data caching in blocking queue and workers are duplicated, remove blocking queue for saving memory
6. put whole pipeline in python side to make `_send_idx` `_rcvd_idx` and `_batch_outstanding` state more robust

### Speed test

testing based on `hapi` code and 4 * V100

| Model | batch size/GPU | worker num/GPU | develop speed | this PR speed | speed up ratio |
| :--- | :---: |   :---: |   :---: |   :---: |   :---: | 
| YOLOv3-DarkNet53 | 12 | 4 | 521ms/step | 513ms/step | 1.55% |
| TSM-ResNet50 | 8 | 4 | 276ms/step | 267ms.step |  3.37% |
| ResNet50 | 64 | 4 | 234ms/step | 228ms/step | 2.63% |
| MobileNetV1 | 64 | 4 | 168ms/step | 166ms/step | - |

### Document

### CODE REVIEW

**Review by saxon-zh — DISMISSED** [review]
LGTM

**Review by saxon-zh — DISMISSED** [review]
LGTM

File: python/paddle/fluid/dataloader/collate.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    [batch_filed1, batch_filed2, ...]
+
+    Args:  
+        batch(list of list of numpy array): the batch data, each fields
+              should be a numpy array, each sample should be a list of
+              fileds, and batch should be a list of sample.
+    
+    Returns:
+        a list of numpy array: collated batch
+    """
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/collate.py:56]: maybe assert input type

File: python/paddle/fluid/dataloader/collate.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        batch = np.stack(batch, axis=0)
+        # if CPU-only version, pin_memory will be disabled,
+        # so we compate tensor here
+        if not paddle.is_compiled_with_cuda():
+            batch = _to_tensor(batch)
+        return batch
+    elif isinstance(sample, paddle.Tensor):
+        return layers.stack(batch, axis=0)
+    elif isinstance(sample, (int, float)):
+        batch - np.array(batch)
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/collate.py]: What is the purpose of this line?

File: python/paddle/fluid/dataloader/collate.py (hunk 1 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/collate.py:56]: Done, Thanks!

File: python/paddle/fluid/dataloader/collate.py (hunk 2 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/collate.py]: fixed, and change `test_multiprocess_dataloader_iterable_dataset_split.py` dataset yield `int` data as unittest of this condition

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by saxon-zh — DISMISSED** [review]
LGTM

File: python/paddle/fluid/dataloader/collate.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+try:
+    from collections.abc import Sequence, Mapping
+except:
+    from collections import Sequence, Mapping
+
+
+def _to_tensor(data):
+    if in_dygraph_mode():
+        return paddle.to_tensor(data, place=paddle.CPUPlace())
+    else:
```
Comment (phlrain) [inline python/paddle/fluid/dataloader/collate.py:29]: tensor.set 是一个很低效的操作，有没有测试过静态图的性能变化

**Comment by phlrain** [comment]
这个模块对于动静态图几乎所有模型都会有非常大的影响，是不是应该有更加充分的测试，以及qa帮忙做验证
目前的测试是基于4卡的任务，该pr会不会引起cpu资源消耗增加，在4卡下cpu资源充足，性能有提升，在8卡下会不会有资源不足的问题
由于影响面非常广，建议走下技术评审

File: python/paddle/fluid/dataloader/collate.py (hunk 3 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/collate.py:29]: 之前的实现里面在放到blocking queue之前也是需要通过tensor.set成一个cpu tensor的，也都是在thread里完成的，这部分的开销应该是一致的 https://github.com/PaddlePaddle/Paddle/blob/49e79cad3974bd9738e445e0ec10e6898018a9c6/python/paddle/fluid/dataloader/dataloader_iter.py#L330
目前除了高层API没有其他模型静态图用到多进程DataLoader，我补充一下高层API里的静态图测试数据

**Pull request closed without merging**